File: Data Structures & Algorithms/binary-search/submission-3.cpp
class Solution {
public:
    int search(vector<int>& nums, int target) {
        // Creating up pointers for binary search
        int left = 0, right = nums.size() - 1;

        // Begin our binary search traversal
        while (left <= right) {
            // Calculate our middle index
            int mid = left + (right - left) / 2;

            // Returning mid if it matches the target value
            if (nums[mid] == target) {
                return mid;
            } else if (nums[mid] < target) {
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }

        // In the end, both the left and right pointer will be the index
        // of the peak
        return -1;
    }
};

File: Data Structures & Algorithms/three-integer-sum/submission-3.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        # First, we have to sort the list to
        # use the two pointer techinque
        nums.sort()

        # Creating a final container to hold the final results
        result = []

        # Traverse the entire the list
        for i in range(len(nums)):
            # Skipping any duplicates we encounter
            if i > 0 and nums[i] == nums[i - 1]:
                continue
            
            # Initializing our two pointers
            left, right = i + 1, len(nums) - 1

            # Performing the two pointer approach
            while left < right:
                # Calculating the sum from all of our pointers
                total = nums[i] + nums[left] + nums[right]

                # Record the triplet if our sum is zero
                if total == 0:
                    result.append([nums[i], nums[left], nums[right]])

                    # Adjust pointers as needed
                    left += 1
                    right -= 1

                    while left < right and nums[left] == nums[left - 1]:
                        left += 1

                    while left < right and nums[right] == nums[right + 1]:
                        right -= 1
                else:
                    # Depending on our sum value, we have to
                    # move the corresponding pointer
                    if total < 0:
                        left += 1
                    elif total > 0:
                        right -= 1

        # Returning the final list containing all
        # triplets
        return result

        